notebooker/web/utils.py

```diff
@@ -46,24 +46,128 @@ def get_directory_structure(starting_point: Optional[str] = None) -> Dict[str, U
     """
     Creates a nested dictionary that represents the folder structure of rootdir
     """
+    categorization = current_app.config.get("CATEGORIZATION", False)
     starting_point = starting_point or _get_python_template_dir()
     all_dirs = {}
     rootdir = starting_point.rstrip(os.sep)
     start = rootdir.rfind(os.sep) + 1
-    for path, dirs, files in os.walk(rootdir):
+
+    for path, _, files in os.walk(rootdir):
         if not _valid_dirname(path):
             continue
         folders = path[start:].split(os.sep)
-        subdir = {
-            os.sep.join(folders[1:] + [f.replace(".ipynb", "").replace(".py", "")]): None
-            for f in files
-            if _valid_filename(f)
+        subdir = {}
+        parent = all_dirs
+
+        for f in files:
+            full_path = os.path.join(starting_point, *folders[1:], f)
+            if _valid_filename(f):
+                if categorization:
+                    category = _extract_category(full_path)
+                    if category:
+                        parent.setdefault(rootdir.split(os.sep)[-1], {}).setdefault(category, {})[
+                            os.path.join(*folders[1:], f)
+                        ] = None
+                else:
+                    subdir[os.path.join(*folders[1:], f)] = None
+
+        for folder in folders[:-1]:
+            if folder not in parent:
+                parent[folder] = {}
+            parent = parent[folder]
+        if not categorization:
+            parent[folders[-1]] = subdir
+
+    if categorization:
```

Comment:
I think it would make sense to have a separate function entirely for categorisation which uses the output of `get_directory_structure()` - this addition complicates the function and changes its meaning too much.